Week2

jf8989 opened the pull request.

RubenJTL commented on Memorize/Memorize/Views/MemorizeView.swift:
```diff
@@ -0,0 +1,159 @@
+                    .onTapGesture {
+                        viewModel.choose(card)
+                    }
+                    .allowsHitTesting(viewModel.isTapEnabled)
```
I think the aspect ratio is missing here. Once you add it, you’ll notice that some cards don’t fit in the screen correctly in certain themes.

RubenJTL commented on Memorize/Memorize/Views/MemorizeView.swift:
```diff
@@ -0,0 +1,159 @@
+    // This is the button's action.
+    private func newGameAction() {
+        viewModel.newGame()
+    }
```
I don’t think this additional function is necessary here. You can call the ViewModel action directly in the button’s action.

RubenJTL commented on Memorize/Memorize/ViewModels/MemorizeViewModel.swift:
```diff
@@ -0,0 +1,195 @@
+        rulebook.score
+    }
+
+    @Published var timeRemaining: Int = 120
```
Keep everything organized and move the publishers to the beginning.

RubenJTL replied:
https://www.avanderlee.com/swiftui/mvvm-architectural-coding-pattern-to-structure-views
https://ahmetgeymen.medium.com/how-to-order-property-wrappers-in-swiftui-c4d168d35572

RubenJTL commented on Memorize/Memorize/ViewModels/MemorizeViewModel.swift:
```diff
@@ -0,0 +1,195 @@
+
+    @Published var timeRemaining: Int = 120
+
+    private var timer: Timer?
```
Good research on the timer. But this introduces unnecessary complexity—you can simplify the logic by using https://developer.apple.com/documentation/swiftui/text/init%28timerinterval:pausetime:countsdown:showshours:%29

RubenJTL commented on Memorize/Memorize/ViewModels/MemorizeViewModel.swift:
```diff
@@ -0,0 +1,195 @@
+                self.timer = nil
+                self.isTapEnabled = false  // freeze the game when the time is over
+            }
+        }
```
I noticed that you used the timer to add this new rule, and I like it. However, you should move this logic into its own function.

RubenJTL replied:
Also, make sure to add a deinit because Timer is a class and could potentially cause memory leaks if not invalidated properly.

RubenJTL commented on Memorize/Memorize/ViewModels/MemorizeViewModel.swift:
```diff
@@ -0,0 +1,195 @@
+    )  // reference to the Rulebook struct in the model
+
+    var safeNumberOfPairs: Int {
+        guard let theme = selectedTheme else { return 8 }
+
+        let emojiCount = theme.emojis.count
+
+        if let pairs = theme.numberOfPairs {
+            if pairs <= emojiCount && pairs > 4 {
+                return pairs
+            } else if pairs <= 4 {
+                return 8
+            } else if pairs > emojiCount {
+                return 9
+            } else {
+                return 7
+            }
+        } else {
+            let minPairs = min(6, 9)
+            let maxPairs = 10
+            if minPairs > maxPairs { return maxPairs }
+            return Int.random(in: minPairs...maxPairs)
+        }
```
This is really confusing and tends to cause errors. Try to simplify the logic and prevent possible issues. For example, what happens if emojiCount is 0 or smaller than the required pairs?

RubenJTL replied:
There is logic that isn’t needed, such as `let minPairs = min(6, 9)` since this will always return 6. You can simplify or remove this part to avoid unnecessary complexity.

RubenJTL commented on Memorize/Memorize/ViewModels/MemorizeViewModel.swift:
```diff
@@ -0,0 +1,195 @@
+
+    var themeGradientColor: LinearGradient? {
+        guard let color1Name = selectedTheme?.color,
+            let color2Name = selectedTheme?.colorG
```
indentation

RubenJTL commented on Memorize/Memorize/ViewModels/MemorizeViewModel.swift:
```diff
@@ -0,0 +1,195 @@
+            timeRemaining -= 3
+            isTapEnabled = false  // block any taps for now
+
+            DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
```
Good find. In a future lesson, we’ll talk about Publishers, and you’ll notice that you can simplify this logic as well.

RubenJTL replied:
The score calculation logic doesn’t seem to be directly related to the assignment and also tends to cause errors with the timer <img width="352" height="707" alt="image" src="https://github.com/user-attachments/assets/a029fb52-6763-4059-b344-dd0bcbd6a799" />

RubenJTL commented on Memorize/Memorize/Model/RulebookModel.swift:
```diff
@@ -0,0 +1,85 @@
+                // MISMATCH
+                // Score -1 for each card previously seen.  Tags both cards as been seen before, and for each missmatch in this case, the user gets -1 score.
+                if cards[firstIndex].hasBeenSeen { score -= 4 }  // checks if this is the first time they've been seen.  If not, -1 points for the user.
+                if cards[secondIndex].hasBeenSeen { score -= 4 }
```
I think the comment is not related to what the code actually does.

RubenJTL commented on Memorize/Memorize/Views/MemorizeView.swift:
```diff
@@ -0,0 +1,159 @@
+                    .opacity(0.25)
+                    .offset(y: 26)
+            )
+            .animation(
+                .spring(
+                    response: 0.45,
+                    dampingFraction: 0.55,
+                    blendDuration: 0.7
+                ),
+                value: viewModel.themeName
+            )
```
I think spring is not the best animation for this case
https://medium.com/@arunzzrip/swiftui-animations-the-basics-of-animations-and-transitions-a6837b7ed5e2

RubenJTL commented on Memorize/Memorize/Views/MemorizeView.swift:
```diff
@@ -0,0 +1,159 @@
+struct MemorizeView: View {
+    @Environment(\.colorScheme) var colorScheme  // This allows me to decide what happens when the Color Scheme is "dark", preventing color-masking.
+
+    @ObservedObject var viewModel = MemorizeViewModel()  // @ObservedObject means SwiftUI watches for changes and updates the View automatically.
```
When you use @ObservedObject, it’s because you plan to instantiate the class in a parent view. This is important because every time the parent view is re-rendered, the child view (e.g., MemorizeView) will create a new ViewModel. Try to read more about this behavior.
https://www.avanderlee.com/swiftui/stateobject-observedobject-differences/

RubenJTL commented on Memorize/Memorize/Views/MemorizeView.swift:
```diff
@@ -0,0 +1,159 @@
+                    .fontDesign(.serif)
+            }
+        }
+        .opacity(card.isMatched ? 0 : 1).allowsHitTesting(!card.isMatched)
```
Move the modifier to the next line.

RubenJTL commented:
Good improvements in your commits! Try to reduce the number of comments in the code, as they’re not always useful. Keep in mind that when something is well implemented, additional comments are often unnecessary.

Also, stay focused on the assignment requirements. I noticed that some parts still need refinement, such as the score calculation logic (how points are added or deducted) and the correct aspect ratio for the cards.

jf8989 commented:
Hi Ruben,

Thanks for your thorough review and suggestions!
I’ve addressed all your comments:

Refined the score calculation logic to align with assignment requirements.

Fixed the card aspect ratio and updated the layout for better consistency.

Moved @Published properties to the top of the ViewModel and reorganized publishers.

Modularized the timer logic and added a deinit for proper cleanup.

Simplified the number of pairs logic for clarity and edge cases.

Made other minor fixes based on your feedback (indentation, code style, etc.).

Let me know if you have any further suggestions or if there’s anything else you’d like me to clarify before Monday.

Thanks again for your feedback!
